jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java

```diff
@@ -50,7 +50,7 @@ public class PathResource extends Resource
     private final Path path;
     private final URI uri;
     private boolean alias = false;
-    private Path targetPath;
+    private Path canonicalPath;
```

Comment:
Why "canonical"?
This is confusing to me.

In symlink terms, this PathResource is the requested resource.
The symlink target is the destination.

In windows extensionless terms, this PathResource is the requested resource.
The target is the actual file that would be opened in the FileSystem and FileStore objects.

In unicode normalization terms, this PathResource is the requested resource.
The target is the actual file that would be opened in the FileSystem and FileStore objects.

"canonical" is _not a term_ used by Path, FileSystem, or FileStore.
"Target" is, especially so with `toRealPath()` usage.

"canonical" is for `File` or `URI`, which `PathResource` and especially this field is not.

Reply:
Several reasons:
 + "Target" implies that it is a symlink that has a target.    Hopefully most PathResources will not be symlinks.
 + There are other types of alias other than symlink (case insensitive, 8.3 name, normalization). Target means nothing for these type of alias.
 + The changes here have this field being non null for all resources, i.e. even a non aliased resource will have this field set to the same value as path.  This is definitely not a target.
 + The very definition of an alias is when the passed path does not equal the canonical path (or passed URI does not equal the canonical URI), thus we get the good code `aliased = !this.path.equals(this.canonicalPath)`

I was happy with the previous name of `getAliased` which I don't think should have been changed to the symlink only `getTarget`.  But now that it is always non null, `getCanonical` is really what it is.

Reply:
More over, "canonical" is definitely a term used in relation to the File system.  In Jetty 6,7,8,9,10,11 the definition of an alias has been if `file.getCanonicalFile() != file`.  Using the name canonical here is a good link back to previous versions so we can see that we have the same meaning for what an alias is.

Reply:
@gregw after recent changes it is not always true that `getCanonical` is always not null.
It is now null in the case where the file does not exist, or with a broken simlink, but if it returns null the resource is classified as an alias.

So `Resource.getCanonicalURI() == null` is now not equivalent to `isAlias()` like it was before this PR.
And if `Resource.getCanonicalURI() == null` then `Resource.exists()` is always false.

Reply:
@lachlan-roberts that doesn't make sense?  How can `resource.getCanonicalURI() == null` be equiv to isAlias? Do you mean  `resource.getCanonicalURI() != null` is equivalent to `isAlias==true`?
Does `getCanonicalURI()` ever return non-null for a non-aliased resource?   If not, then can we rename it back to `getAlias` or `getAliasedURI`?

Reply:
> How can resource.getCanonicalURI() == null be equiv to isAlias?

I meant `resource.getCanonicalURI() != null` is NOT equivalent to `isAlias==true`.
Before this PR it was used as an equivalent check.

> Does getCanonicalURI() ever return non-null for a non-aliased resource?

Yes as long as the resource exists then `getCanonicalURI()` will be non-null. If it doesn't exist we say the resource is an alias and `getCanonicalURI()` is null because we couldn't resolve it.

Reply:
OK, so "Canonical" still works.  It is the correct representation of that resource... which is null for a non existent resource.

Reply:
"Canonical" is an old term that no longer has relevance.

Note that "target" is the new term, as it refers to both "symlink" and "as storage preferred name" (eg: unicode normalization, case insensitive, escaped names, etc)

The techniques were to follow every target (symlink or as storage preferred name), in every component of the given name recursively, and ____

1. all but the last component must exist (and be accessible by user).
2. all components must exist (and be accessible by user).
3. without requirements of component existence or accessibility.

When java introduced `File.getCanonicalPath()` (in Java 1.1 back in 1997) and `File.getCanonicalFile()` (in Java 1.2 back in 1998) is actually a forced technique 3, which was fine at the time, but is now at odds with modern FileSystems that do not support that technique.
Tooling and other APIs was rampant for how to "canonicalize" a file system object too.
But that's not an API or tooling concern or technique, that's a FileSystem concern (how to get the real file system object name).

So things started to change around 2011.

I'll pick 2 notable events (out of many)

Back in 2012.
The gnu coreutils dropped "canonicalize" as a description of behaviors for filesystems when they moved from `readlink` (now deprecated) to `realpath` (the recommended approach).
You can still find "canonicalize" in the modern gnu coreutils documentation, but it's now historical and refers to in API techniques for resolution, not how the filesystem actually works. (if you look at an old version of coreutils documentation from before the introduction of `realpath` and you'll find "canonical" this and that everywhere in the documentation)

Back in late 2011.
The `File.getCanonicalFile()` / `File.getCanonicalPath()` APIs around Java 1.7 were updated, and is not a recommended API anymore, as it doesn't convey the truth of how FileSystems work anymore. (both of those APIs document and call out the existence of `Path.toRealPath()` now)

Java introduced the `Path` APIs in Java 1.7, and it was following the industry trend for trusting the filesystem, and not doing everything at the API layer.
The entire concept of canonicalization was updated to be "real path" now (as it referred to both symlink and "as storage preferred name" as well).
The default behavior of `UnixPath` (and `WindowsPath`) for `toRealPath()` is to defer to the FileSystem behavior (you can override this by using the option to not following links, and then it falls back to in-API behavior technique 2 in the above list).

In fact you cannot find "canonical" as a term in the javadoc on `java.nio.file.*` now.

```
[joakim@hyperion jdk-17-srcs]$ grep -RHin "canonical" java.base/java/io/* | wc -l
79
[joakim@hyperion jdk-17-srcs]$ grep -RHin "canonical" java.base/java/nio/file/* | wc -l
0
```

Reply:
Don't introduce old, now-deprecated, terms.
Keep "target" as it applies to the modern era, not the past.

Reply:
"Canonical" still is an english word with meaning.  But `getRealURI()` works for me.   "Target" is still just wrong in many circumstances.

Reply:
Note also that we have recently added `getCanonicalPath` to `HttpURI`.  So we do have historical and current ties with the concept of canonical URIs and Paths. 

Moreover, one of the whole points of the `Resource` abstraction is that we simply cannot adopt an approach of "trusting the Filesystem".   It is precisely because of Filesystems forgiving nature that we have problems with aliasing that result in security issues.   So we explicitly do not defer resolution to the Filesystem.  We precisely work out a canonical URI, which we then wish to apply to a base Resource in a process that will reliably tell us if we have requested the origin/canonical/real resource or an alias of that.  

So we simply cannot "defer to the FileSystem" or JVM methods.   We have tried that several times before and every time that has resulted in security problems.  We have to stop repeating that mistake over and over.   

So we have three choices:

1. Go back to the `getAlias` or `getAliased` style names.   However, I think that @lachlan-roberts has expressed a preference for the ability to get a Path/URI that is non-null (if resource exists) regardless of if it is an alias or not.
2. Live with `getCanonicalURI` due to it's historic and current links with the process we have for detecting and handling aliases.
3. Find a new name ("Target" is not it) that can be applied to both `HttpURI` and `Resource`.   I could live with `getRealURI()`

@lachlan-roberts your call as to which if the above you go with.

Reply:
Canonical makes sense for URI, not for anything related to FileSystems.